test/end2end_test.go

```diff
@@ -107,7 +108,7 @@ func (s *testServer) UnaryCall(ctx context.Context, in *testpb.SimpleRequest) (*
 	md, ok := metadata.FromContext(ctx)
 	if ok {
 		if err := grpc.SendHeader(ctx, md); err != nil {
-			grpclog.Fatalf("grpc.SendHeader(%v, %v) = %v, want %v", ctx, md, err, nil)
+			return nil, fmt.Errorf("grpc.SendHeader(%v, %v) = %v, want %v", ctx, md, err, nil)
```

Comment:
My point is that there are other places which also log fatal on server side. If you want to do the change, you need to change all of them in one shot, which needs some refactoring (e.g., there is one in newPayload). I think it is probably not worth doing that for now.